runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/fn/control/ProcessRemoteBundleOperation.java

```diff
@@ -100,6 +218,80 @@ public void finish() throws Exception {
       } catch (Exception e) {
         throw new RuntimeException("Failed to finish remote bundle", e);
       }
+
+      try {
+        // In Batch, the input watermark advances from -inf to +inf once all elements are processed.
+        // For timers, this means we only fire them at the end of the bundle.
+        fireTimers();
+      } catch (Exception e) {
+        throw new RuntimeException("Failed to finish remote bundle", e);
+      }
+    }
+  }
+
+  // Recursively fires all scheduled timers from the SDK Harness.
```

Comment:
A pedant might take issue with calling this "recursive"... :-)

Reply:
Ack, changed to be more specific.